Tests/BSONTests/ExtendedJSONConversionTests.swift

```diff
@@ -4,7 +4,75 @@ import Nimble
 import NIO
 import XCTest
 
+extension CodingUserInfoKey {
+    static let barInfo = CodingUserInfoKey(rawValue: "bar")!
+}
+
 open class ExtendedJSONConversionTestCase: BSONTestCase {
+    struct Test: Codable, Equatable {
+        let x: Bool
+        let y: Int32
+        let z: BSONRegularExpression
+    }
+
+    struct Foo: Codable, Equatable {
+        let foo: Bool
+
+        enum CodingKeys: String, CodingKey {
+            case foo
+        }
+
+        func encode(to encoder: Encoder) throws {
+            let barInfo = encoder.userInfo[.barInfo] as! Bool
+            var container = encoder.container(keyedBy: CodingKeys.self)
+            if barInfo == true {
+                try container.encode(true, forKey: .foo)
+            } else {
+                try container.encode(self.foo, forKey: .foo)
+            }
+        }
+    }
+
+    func testExtendedJSONDecoderAndEncoder() throws {
+        let decoder = ExtendedJSONDecoder()
+        let encoder = ExtendedJSONEncoder()
+        encoder.mode = .canonical
+
+        let regexStr = "{\"$regularExpression\":{\"pattern\":\"p\",\"options\":\"i\"}}"
+        let extJSON = "{\"x\":true,\"y\":{\"$numberInt\":\"5\"},\"z\":\(regexStr)}"
+        let data = extJSON.data(using: .utf8)!
+
+        let regexObj = BSONRegularExpression(pattern: "p", options: "i")
+        let test = Test(x: true, y: 5, z: regexObj)
+
+        let decoded = try decoder.decode(Test.self, from: data)
+        let encoded: Data = try encoder.encode(test)
```

Comment:
can we re-order this test so that it does all of the encoder tests and then all of the decoder tests? As it is right now, theyre intermingled and it's a little hard to follow.

Reply:
done